test/sslapitest.c

```diff
@@ -1964,19 +1964,20 @@ static int hostname_cb(SSL *s, int *al, void *arg)
 
 static const char *servalpn;
 
-static int alpn_select_cb (SSL *ssl, const unsigned char **out, unsigned char *outlen,
-                    const unsigned char *in, unsigned int inlen, void *arg)
+static int alpn_select_cb(SSL *ssl, const unsigned char **out,
+                          unsigned char *outlen, const unsigned char *in,
+                          unsigned int inlen, void *arg)
 {
-    unsigned int i, protlen = 0;
+    unsigned int protlen = 0;
     const unsigned char *prot;
 
-    for (i = 0, prot = in; i < inlen; i += protlen, prot += protlen) {
-        protlen = *(prot++);
```

Comment:
I wonder if adding a simple "i++" here would be better fix.
I for one find it hard to follow the pointer arithmetics in "in + inlen - prot < protlen".

Reply:
So if it's like this: `end = in + inlen` and then `end - prot < protlen`. This should be more comprehensible. The unnecessary `i` is just the reason causing this bug...